sktime/pipeline/pipeline.py

```diff
@@ -0,0 +1,584 @@
+        if list is unnamed, names are generated as names of classes
+        if names are non-unique, `f"_{str(i)}"` is appended to each name string
+            where `i` is the total count of occurrence of a non-unique string
+            inside the list of names leading up to it (inclusive)
+    `add_step(skobject, name, edges, method, **kwargs)` - adds a skobject to the
+        pipeline and setting the name as identifier and the steps specified with
+        edges as input steps (predecessors). Therby the method that should be
+        called can be overriden using the method kwarg. Further provided kwargs
+        are directly provided to the skobject if it is called.
+
+    Parameters
+    ----------
+    param step_informations : what it is, what it does
+
+    Attributes
+    ----------
+    attribute id_to_true_id : a dict with integer keys and values,
+        mapping the python object id to skobject ids.
+    attribute id_to_obj : a dict with integer keys and weak references of
+        skobjects as values. The values are the weak references of the skobjects
+        provided to the `add_step` method. We store the weak references to
+        avoid that the id of the object is reassigned if the user deletes all it
+        references to the object.
+    attribute model_dict : a dict with integer keys and skobject values.
+        This is a mapping of the id of the skobjects provided to `add_step`
+        to the cloned skobject.
+    attribute counter : integer, counts the number of steps in the pipeline.
+    attribute steps : a dict with string keys and step object values.
+        The key is the name that is specified if a skobject is added to the pipeline.
+    attribute kwargs : a dict with str keys and object values. Stores all kwargs
+        that are specified and might be passed to skobjects in the pipeline.
+
+    Examples
+    --------
+    >>> from sktime.classification.distance_based import KNeighborsTimeSeriesClassifier
+    >>> from sktime.datasets import load_arrow_head, load_longley
+    >>> from sktime.forecasting.model_selection import temporal_train_test_split
+    >>> from sktime.forecasting.naive import NaiveForecaster
+    >>> from sktime.pipeline.pipeline import Pipeline
+    >>> from sktime.transformations.compose import Id
+    >>> from sktime.transformations.series.boxcox import BoxCoxTransformer
+    >>> from sktime.transformations.series.exponent import ExponentTransformer
+
+        Example 1: Simple sequential pipeline of transformers using the generalized
+        non-sequential pipeline implementation
+    >>> y, X = load_longley()
+    >>> general_pipeline = Pipeline()
+    >>> for step in [
+    ...     {"skobject": ExponentTransformer(), "name": "exp", "edges": {"X": "X"}},
+    ...     {"skobject": BoxCoxTransformer(), "name": "box", "edges": {"X": "exp"}},
+    ...     ]:
+    ...     general_pipeline = general_pipeline.add_step(**step)
```

Comment:
this seems cumbersome - should it not be possible to simply construct the pipeline with the `step` argument, and have `add_step` as a secondary option?

Reply:
You mean passing the list to the creation of the pipeline?  It is almost possible, I will enable it. Would you then change all tests?

Reply:
yes, that's what I meant. I would still allow both types of syntax though.